src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -325,6 +329,72 @@ EnsureCitusTableCanBeCreated(Oid relationOid)
 }
 
 
+/*
+ * undistribute_table gets a distributed table name and 
+ * udistributes the table.
+ */
+Datum
+undistribute_table(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	Oid relationId = PG_GETARG_OID(0);
+	const char *relationName = get_rel_name(relationId);
+
+	EnsureTableOwner(relationId);
+
+	Relation relation = relation_open(relationId, ExclusiveLock);
+
+	StringInfo query = makeStringInfo();
+
+	SPI_connect();
+
+	List * tableDDLEvents = GetTableDDLEvents(relationId, true);
+	char * createTableQuery = linitial(tableDDLEvents);
```

Comment:
It seems a bit implicit that the first element of `tableDDLEvents` will be `createTableQueryString`.

How about we use a method, something like `PopCreateTableQueryString`. The method can have the same lines as this one so:

```c
/*
 * PopCreateTableQueryString returns the table creation query string from the tableDDLEvents and removes it from the list.
 * Currently it will be the first element of the list.
 */
char* PopCreateTableQueryString(List* tableDDLEvents) {
char * createTableQueryString = linitial(tableDDLEvents);
list_delete_first(tableDDLEvents);
return createTableQueryString;
}
```

I think that this has the advantage that the detail is hidden from the caller now if we use the function and there is only one place to update how we get the table creation query string from the ddl events. 

And to be more robust we can iterate the ddl event list to find the create table query string in this method so that even if it is not the first one we still return it correctly, but I think it is not that necessary for now.

What do you think about the above method?

Reply:
I agree that this is a bit hacky. Reading the code it even seems like the current solution is incorrect when it involves a foreign table. Since then the first querystring is not actually the create table:
https://github.com/citusdata/citus/blob/0259815d3ab96e41be7ec1e00e959411097a0619/src/backend/distributed/operations/node_protocol.c#L598-L615

This would also be the case for the approach suggested by talha.

I think it makes more sense to call `GetTableCreationCommands` commands manually and then strip that amount of elements from the result of `GetTableDDLEvents`. If you want to make it even nicer you could add a `GetTableDDLEvents` version that doesn't call `GetTableCreationCommands` at all. But I'm not sure if that's worth the effort and the clutter TBH.